File: Algo_Study_Plan/Binary_Search/34_Find_First_and_Last_Postition_of_Element_in_Sorted_Array/find_first_and_last_position_of_element_in_sorted_array.py
#code by dev-josh on leetcode. Although not the most efficient I found this approach to be quite intelegent.
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        
        def search(x):
            lo, hi = 0, len(nums)           
            while lo < hi:
                mid = (lo + hi) // 2
                if nums[mid] < x:
                    lo = mid+1
                else:
                    hi = mid                    
            return lo
        
        lo = search(target)
        hi = search(target+1)-1
        
        if lo <= hi:
            return [lo, hi]
                
        return [-1, -1]


# Explaination regarding cases from simite200 on LeetCode

# search function
#    find the lowest index of target num in nums
#         if there is target num in nums
#             lo = lowest index of target
#             hi = lowest index of target
#         if there is no target num in nums
#             if there is higher num than target in nums
#                 lo = lowest index of the first higher num than target in nums
#                 hi = lowest index of the first higher num than target in nums
#             if there is no higher num than target in nums
#                 lo = len(nums)
#                 hi = len(nums)
# after lo = search(target), hi = search(target+1) -1
# if there is more than one target num
#     lo = lowest index of target
#     hi = (lowest index of the first higher num than target) -1 ( highest index of target)
#     then, lo < high
# elif there is only one target num
#     lo == high
# else
#     if there is only higher number than target in nums
#         lo = lowest index of the first higher num than target(at least target+1 or higher)
#         hi = lowest index of the (target+1) - 1
#         then, lo > high
#     if there is only lower number than target in nums
#         lo = len(nums)
#         hi = len(nums)-1
#         then, lo > high

File: Algo_Study_Plan/Binary_Search/34_Find_First_and_Last_Postition_of_Element_in_Sorted_Array/find_first_and_last_position_of_element_in_sorted_array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int first = 0;
        int last = nums.length -1;
        int[] result = {-1, -1};
        
        //Edge cases
        if(nums.length == 0) {
            return result;
        } 
        
        //Find the first position
        while(first < last) {
            int mid = (first + last) / 2;
            if(nums[mid] < target) {
                first = mid +1;
            }
            else {
                last = mid;
            }
        }
        if(nums[first] == target) {
            result[0] = first;
            }
        else {
           return result;
        }
        
        //Find the last position
        //Reset last, first does not need to be reset
        last = nums.length - 1;
        while(first < last){
            int mid = ((first + last)/2) + 1; //add 1 to mid to keep start from getting stuck
            if(nums[mid] > target) {
                last = mid - 1;
            }
            else {
                first = mid;
            }       
        }
        result[1] = last;
        return result;    
    }
}